src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/UpsertProviderTLevelDefinitionsHandler.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System.Data.SqlClient;
+using System.Linq;
+using System.Threading.Tasks;
+using Dapper;
+using Dfc.CourseDirectory.Core.DataStore.Sql.Queries;
+using OneOf.Types;
+
+namespace Dfc.CourseDirectory.Core.DataStore.Sql.QueryHandlers
+{
+    public class UpsertProviderTLevelDefinitionsHandler : ISqlQueryHandler<UpsertProviderTLevelDefinitions, None>
+    {
+        public async Task<None> Execute(SqlTransaction transaction, UpsertProviderTLevelDefinitions query)
+        {
+            const string createTableSql = @"
+CREATE TABLE #TLevelDefinitionIds (
+    TLevelDefinitionId UNIQUEIDENTIFIER
+)";
+
+            await transaction.Connection.ExecuteAsync(createTableSql, transaction: transaction);
+
+            await BulkCopyHelper.WriteRecords(
+                query.TLevelDefinitionIds.Distinct().Select(id => new { TLevelDefinitionId = id }),
+                tableName: "#TLevelDefinitionIds",
+                transaction);
+
+            const string sql = @"
+MERGE Pttcd.ProviderTLevelDefinitions AS target
+USING (
+    SELECT
+        TLevelDefinitionId
+    FROM #TLevelDefinitionIds
+) AS source
+ON target.ProviderId = @ProviderId
+AND target.TLevelDefinitionId = source.TLevelDefinitionId
+WHEN NOT MATCHED THEN
+    INSERT (
+        ProviderId,
+        TLevelDefinitionId
+    ) VALUES (
+        @ProviderId,
+        source.TLevelDefinitionId
+    )
+WHEN NOT MATCHED BY SOURCE THEN
```

Comment:
```suggestion
WHEN NOT MATCHED BY SOURCE AND target.ProviderId = @ProviderId THEN
```

Reply:
Any particular reason for this? This condition is given in the main using statement and removing it from there would affect `WHEN NOT MATCHED`.

Reply:
Pretty sure it will remove any entries for other providers as written

Reply:
That's surprising, good spot